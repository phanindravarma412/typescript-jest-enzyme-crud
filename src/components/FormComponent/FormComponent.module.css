.form {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin-top: 20px;
}

.input-parent {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.input-parent label {
  margin-bottom: 10px;
}

.input-parent input {
  width: 100%;
  height: 40px;
  border-radius: 5px;
  border: 1px solid #ccc;
  padding: 0 10px;
}

.form button {
  width: 16%;
  height: 40px;
  border-radius: 5px;
  border: none;
  background-color: rgb(18, 223, 18);
  color: #fff;
  font-weight: bold;
  font-size: 16px;
  cursor: pointer;
  margin-top: 24px;
}
