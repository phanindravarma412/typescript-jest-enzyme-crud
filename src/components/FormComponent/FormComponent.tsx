import React from "react";

import styles from "./FormComponent.module.css";
import { UserModal } from "../../models/UserModal";

interface formProps {
  usersData: UserModal[];
  setUsersData: React.Dispatch<React.SetStateAction<UserModal[]>>;
  editedUser: UserModal | undefined;
}

const FormComponent: React.FunctionComponent<formProps> = ({
  setUsersData,
  usersData,
  editedUser,
}) => {
  const [userFormData, setUserFormData] = React.useState<UserModal>({
    id: "",
    name: "",
    email: "",
    phone: "",
  });

  const handleChange = (e: React.SyntheticEvent) => {
    editedUser
      ? setUserFormData({
          ...userFormData,
          [(e.target as HTMLInputElement).name]: (e.target as HTMLInputElement)
            .value,
        })
      : setUserFormData({
          ...userFormData,
          id: `${Math.floor(Math.random() * 100 + 1)}`,
          [(e.target as HTMLInputElement).name]: (e.target as HTMLInputElement)
            .value,
        });
  };

  const formSubmit = () => {
    setUsersData([...usersData, userFormData]);
  };

  const formUpdate = () => {
    console.log(userFormData);
    setUsersData(
      usersData.map((user) =>
        editedUser?.id === user.id
          ? { ...userFormData, user: userFormData }
          : user
      )
    );
  };

  return (
    <>
      <div className={styles.form}>
        <div className={styles["input-parent"]}>
          <label>User Name</label>
          <input
            type="text"
            value={editedUser?.name && userFormData.name}
            test-id="name"
            onChange={handleChange}
            name="name"
          />
        </div>
        <div className={styles["input-parent"]}>
          <label>Email</label>
          <input
            type="email"
            test-id="email"
            value={editedUser?.email && userFormData.email}
            onChange={handleChange}
            name="email"
          />
        </div>
        <div className={styles["input-parent"]}>
          <label>Phone</label>
          <input
            type="number"
            test-id="phone"
            value={editedUser?.phone && userFormData.phone}
            onChange={handleChange}
            name="phone"
          />
        </div>

        <button
          test-id={editedUser ? "updateBtn" : "submitBtn"}
          onClick={editedUser ? formUpdate : formSubmit}
        >
          {editedUser ? "update" : "submit"}
        </button>
      </div>
    </>
  );
};

export default FormComponent;
