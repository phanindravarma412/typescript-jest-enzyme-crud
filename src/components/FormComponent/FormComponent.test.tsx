import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import React from "react";

import FormComponent from "./FormComponent";

Enzyme.configure({ adapter: new Adapter() });

describe("User form component", () => {
  const defaultProps = {
    usersData: [],
    setUsersData: () => {},
    setEditedUser: () => {},
    editedUser: undefined,
    isEdit: false,
  };

  const wrapper = shallow(<FormComponent {...defaultProps} />);


  test("is form is rendering or not render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("input change handler check", () => {
    const onChangeHandle = (id: string, value: any) => {
      const inputField = wrapper.find(`input[test-id="${id}"]`);

      return inputField.simulate("change", { target: { value } });
    };

    onChangeHandle("name", "phani");

    onChangeHandle("email", "phani@gmail.com");

    onChangeHandle("phone", "1234567890");
  });

  test("handle submit", () => {
      const submitBtn = wrapper.find(`[test-id="submitBtn"]`);
      submitBtn.simulate("click");
  })
});

describe("isedit true", () => {
  const user = {
    id: "1",
    name: "phani",
    email: "phani@gmail.com",
    phone: "1234567890",
  };
  test("handle update", () => {
    const defaultProps = {
      usersData: [],
      setUsersData: () => {},
      setEditedUser: () => {},
      editedUser: user,
    };
    const wrapper = shallow(<FormComponent {...defaultProps} />);
    const updateBtn = wrapper.find(`[test-id="updateBtn"]`);
    updateBtn.simulate("click");

    console.log(updateBtn.debug);
  });
});

describe("form component", () => {
  const user1 = {
    id: "1",
    name: "phani",
    email: "phani@gmail.com",
    phone: "1234567890",
  };

  const user2 = {
    id: "2",
    name: "phanindra",
    email: "phanindra@gmail.com",
    phone: "666667890",
  };

  const editedUser = {
    id: "1",
    name: "varma",
    email: "varma@gmail.com",
    phone: "555555555",
  };

  const usersData = [user1, user2];
  const defaultProps = {
    usersData: usersData,
    setUsersData: () => {},
    setEditedUser: () => {},
    editedUser: editedUser,
    isEdit: false,
  };

  const wrapper = shallow(<FormComponent {...defaultProps} />);

  test("is form component was rendered or not", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("handle change", () => {
    const onChangeHandle = (id: string, value: any) => {
      const getChangingInput = wrapper.find(`input[test-id="${id}"]`);
      return getChangingInput.simulate("change", { target: { value } });
    };

    onChangeHandle("name", "varma");
    onChangeHandle("email", "varma@gmail.com");
    onChangeHandle("phone", "555555555");
  });

  const updateBtn = wrapper.find(`[test-id="updateBtn"]`);
  updateBtn.simulate("click");
});
