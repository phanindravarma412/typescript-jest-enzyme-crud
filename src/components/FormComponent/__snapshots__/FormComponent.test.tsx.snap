// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User form component is form is rendering or not render 1`] = `ShallowWrapper {}`;

exports[`form component is form component was rendered or not 1`] = `ShallowWrapper {}`;
